### PR TITLE
GDScript style guide

### PR DESCRIPTION
**Pull request by cbscribe**
Goal: provide a centralized location for "good" GDScript coding practices.

Should be used for all code samples in documentation (mostly is already, but we should have a place to point new contributors).

### CODE REVIEW

File: learning/scripting/gdscript/gdscript_styleguide.rst [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+This styleguide lists conventions to write elegant GDScript. The goal is
+to encourage writing clean, readable code and promote consistency across
+projects, discussions, and tutorials. Hopefully, this will also
+encourage development of auto-formatting tools.
+
+Since GDScript is close to Python, this guide is inspired by Python's
+`PEP 8 <https://www.python.org/dev/peps/pep-0008/>`__ programming
+styleguide.
+
+*Note:* Godot's built-in script editor uses a lot of these conventions
```
Comment (NathanLovato) [inline learning/scripting/gdscript/gdscript_styleguide.rst]: Use the docs' note block
`.. note::`
If you keep it with multiple lines you'll have to create a block with 4 spaces as indent
```
.. note::

    Godot's built-in script editor uses a lot of these conventions
    by default. Let it help you.
```

Comment (NathanLovato) [reply learning/scripting/gdscript/gdscript_styleguide.rst]: Aside from that I think it's good

Comment (cbscribe) [reply learning/scripting/gdscript/gdscript_styleguide.rst]: Good one - I did that in my tutorial, but hadn't gone back and added them to this one

File: learning/scripting/gdscript/gdscript_styleguide.rst [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+**Bad**
+
+::
+    for i in range(10):
+      print("hello")
+      
+    for i in range(10):
+            print("hello")
+
+Use at least 2 indent levels to distinguish continuation lines from
```
Comment (akien-mga) [inline learning/scripting/gdscript/gdscript_styleguide.rst]: I would remove "at least".

File: learning/scripting/gdscript/gdscript_styleguide.rst [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+        Tween.TRANS_QUAD, Tween.EASE_OUT)
+
+Blank lines
+~~~~~~~~~~~
+
+Surround functions and class definitions by a blank line.
+
+Use one blank line inside functions to separate logical sections.
+
+One Statement per line
```
Comment (akien-mga) [inline learning/scripting/gdscript/gdscript_styleguide.rst]: Inconsistent capitalization ;) ("Statement" and "line")

File: learning/scripting/gdscript/gdscript_styleguide.rst [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+
+**Never!**
+
+::
+
+    x        = 100
+    y        = 100
+    velocity = 500
+
+**Exceptions**
```
Comment (akien-mga) [inline learning/scripting/gdscript/gdscript_styleguide.rst]: Not sure about that one. "When using multiple operators in an expression" is not clear. Does that mean that we should write `c = a * a` but `c = a*a + b*b`? That's not consistent.

It should be either space around all operators, or around all operators but `*` and `/`. It's up for discussion, but I guess the former is easier to explain and enforce (and still looks fine).

I tended to recommend the latter (no space around `*`), but consider something like: `position += delta*direction*speed`. That's not readable. `position += delta * direction * speed` is much better.

Comment (cbscribe) [reply learning/scripting/gdscript/gdscript_styleguide.rst]: Yeah, this example came from PEP8. I get the intent - the thinking was probably something around order of operations. I can see how `c = a * a + b * b` requires more thought than `c = a*a + b*b`, but it's probably too fuzzy a distinction to put in a style guide. I'll just take it out.

File: learning/scripting/gdscript/gdscript_styleguide.rst [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+**Good**
+
+::
+
+    for i in range(10):
+        print("hello")
+
+**Bad**
+
+::
```
Comment (pwab) [inline learning/scripting/gdscript/gdscript_styleguide.rst]: Missing linebreak so the code block doesn't render

Comment (cbscribe) [reply learning/scripting/gdscript/gdscript_styleguide.rst]: Fixed.

**Pull request merged**